### ISSUE TITLE
Python SDK: Investigate and make it as easy as possible to pass a different configuration provider

### ISSUE THREAD

**Issue by josephaxisa** [looker-open-source/sdk-codegen#140]
Currently this can be done by passing a custom configuration provider to the generated methods constructor, like so:

```
from looker_sdk.rtl import auth_sesssion, requests_transport, serialize
from looker_sdk.sdk import methods

settings = custom_configuration_provider()
transport = requests_transport.RequestsTransport.configure(settings)

sdk = methods.LookerSDK(
        auth_session.AuthSession(settings, transport, serialize.deserialize)
        serialize.deserialize,
        serialize.serialize,
        transport,    
)
```

The only issue is that on [L122-130](https://github.com/looker-open-source/sdk-codegen/blob/master/python/looker_sdk/rtl/auth_session.py#L122-L130) in `AuthSession` we are relying on a `read_ini` function which needs to be present in `ApiSettings`, although it can be reading the credentials from wherever, such as a database. This can be made a bit more generic to fit whatever configuration provider is being used.

#137 will eventually describe the example given above in more detail.

**Comment by joeldodge79** [comment]
@josephaxisa - @jkaster and I discussed this and we like the idea of using an "Interface" (Interface in typescript). Let's use [Protocols](https://www.python.org/dev/peps/pep-0544/) which can be installed in 3.7 via [pypi](https://pypi.org/project/typing-extensions/)

adding something like this to transport.py

```python
class PTransportSettings(Protocol):
    base_url: str
    api_version: str
    verify_ssl: bool
    timeout: int
    headers: Optional[MutableMapping[str, str]]
```

and update `Transport.configure` and `Transport.request` signatures to reference the new `PTransportSettings` type. Our default implementation will continue to use `TransportSettings`

then in auth_session.py add something like

```python
class PAuthSettings(transport.PTransportSettings):
    def get_client_id(self) -> str:
    ...

    def get_client_secret(self) -> str:
    ...
```

Then we update `ApiSettings` to implement `get_client_id` and `get_client_secret` and update `AuthSession.__init__` to take a `PAuthSettings`. It might be cleaner to move `ApiSettings` into `auth_session.py and name it `AuthSettings` instead?

@jkaster what do you think of that interface?

**Comment by jkaster** [comment]
the other "configuration providers" interfaces add to transport settings:
```typescript
isConfigured(): boolean // returns false if required configuration of base_url and api_version are missing 
readConfig() -> Map<String,String> // Reads all configuration values into key/value pairs
```
where `client_id` and `client_secret` would be two of the key/value entries in `readConfig()`

**Comment by jkaster** [comment]
Closed by PR #149
